http/handler.go

```diff
@@ -730,11 +738,74 @@ func parseQuery(values url.Values) map[string]interface{} {
 	return nil
 }
 
+type ctxKeyMaxRequestJsonComplexity struct{}
+
+func maximumJsonTokensFromContext(ctx context.Context) int64 {
+	maxJsonTokens := ctx.Value(ctxKeyMaxRequestJsonComplexity{})
+	if maxJsonTokens == nil {
+		return 0
+	}
+	return maxJsonTokens.(int64)
+}
+
+func addMaximumJsonTokensToContext(ctx context.Context, limit int64) context.Context {
+	return context.WithValue(ctx, ctxKeyMaxRequestJsonComplexity{}, limit)
+}
+
 func parseJSONRequest(r *http.Request, w http.ResponseWriter, out interface{}) (io.ReadCloser, error) {
-	// Limit the maximum number of bytes to MaxRequestSize to protect
-	// against an indefinite amount of data being read.
-	reader := r.Body
-	err := jsonutil.DecodeJSONFromReader(reader, out)
+	ctx := r.Context()
+	maxJsonTokens := maximumJsonTokensFromContext(ctx)
+
+	if maxJsonTokens <= 0 {
+		err := jsonutil.DecodeJSONFromReader(r.Body, out)
+		if err != nil && err != io.EOF {
+			return nil, fmt.Errorf("failed to parse JSON input: %w", err)
+		}
+		return nil, err
+	}
+
+	reader, ok := r.Body.(io.ReadSeeker)
+	if !ok {
+		body, err := io.ReadAll(r.Body)
+		reader = bytes.NewReader(body)
+		if err != nil {
+			return nil, fmt.Errorf("failed to read JSON input: %w", err)
+		}
+	}
+
+	pos, err := reader.Seek(0, io.SeekCurrent)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read JSON input: %w", err)
+	}
+
+	dec := json.NewDecoder(reader)
+
+	var tokenCount int64
+	for {
+		_, err := dec.Token()
+		if err != nil {
+			if errors.Is(err, io.EOF) {
+				break
+			}
+			return nil, err
+		}
+		tokenCount++
+		if tokenCount > maxJsonTokens {
+			return nil, errors.New("failed to parse JSON input: too many tokens")
+		}
+	}
```

Comment:
I might try a different approach to this, but if we want to merge this now, I'm fine with it. 

I'm thinking about also adding a `max_string_count` in addition to `max_token_count`, as each string incurs another HMAC invocation, which brings more overhead. Fewer HMACs the better, IMHO.